docs/generic-templates.md

```diff
@@ -0,0 +1,173 @@
+1. `LfTemplate` indicates an "actual template" which will cause a template definition in DAML-LF. Our example above will yield `instance LfTemplate I`. The `LfTemplate` type class serves the same purpose as the current `Template` type class but is not intended to be used in DAML. Instances of `LfTemplate` are generated for generic template instantiations and non-generic template definitions.
+1. `Template` indicates a "logical template" which exists only in DAML but not in DAML-LF. Our example above will yield `instance D a => Template (T a)`. Instances of `Template` are generated for all template definitions, generic and non-generic ones.
+1. `Creatable` links an instance of `Template` with an instance of `LfTemplate` (at the value level). Our example above will yield `instance Creatable (T S)`. Instances of `Creatable` are generated for generic template instantiations and non-generic template definitions.
+
+The actual definitions of the three type classes are as follows:
+```haskell
+class LfTemplate x where
+  lfSignatory : t -> [Party]
+  lfCreate : t -> Update (ContractId t)
+  lfCreate = magic @"create"
+  lfFetch : ContractId t -> Update t
+  lfFetch = magic @"fetch"
+
+class Template t where
+  signatory : t -> [Party]
+
+class Template t => Creatable t where
+  create : t -> Update (ContractId t)
+  fetch : ContractId t -> Update t
+```
+
+Similarly, we split the `Choice` type class into `LfChoice`, `Choice` and `Exercisable`:
+```haskell
+class LfTemplate t => LfChoice t c r | t c -> r where
+  lfController : t -> c -> [Party]
+  lfAction : ContractId t -> t -> c -> Update r
+  lfExercise : ContractId t -> c -> Update r
+  lfExercise = magic @"exercise"
+
+class Template t => Choice t c r | t c -> r where
+  controller : t -> c -> [Party]
+  action : ContractId t -> t -> c -> Update r
+
+class (Creatable t, Choice t c r) => Exercisable t c r | t c -> r where
+  exercise : ContractId t -> c -> Update r
+```
+
+## Desugaring generic template definitions
+
+Desugarting the generic template definition of `T` from above will generate the following data definitions type class instances:
+```haskell
+data T a = T with f_1 : U_1; ...
+
+instance D a => Template (T a)
+  signatory this@T{..} = sign
+
+
+data C a = C with g_1 : V_1; ...
+
+instance D a => Choice (T a) (C a) R where
+  controller this@T{..} arg@C{..} = ctrl
+  action self this@T{..} arg@C{..} = act
+```
+Extending this desugaring to more than one type parameter is trivial.
+
+## Desugaring generic template instantiations
+
+Desugaring the generic template instantion `I = T S` from above yields:
+```haskell
+newtype I = MkI with unI : T S
+
+instance LfTemplate I where
+  lfSignatory = signatory . unI
+
+instance Creatable (T S) where
+  create t = fmap coerceContractId . lfCreate @I . MkI
+  fetch = fmap unI . genFetch @I . coerceContractId
+
+
+instance LfChoice I (C S) R[S/a] where
```

Comment:
I presume the notation `R[S/a]` is indicating an algorithm in the desugaring that substitutes `S` for `a` in `R`.

Reply:
Yes, that's substitution. GHC already provides functions for this.

Reply:
For my interest, are these familiar to you? Can you point me at them please?